action.yml

```diff
@@ -0,0 +1,232 @@
+name: Run tests
+description: Run tests for a single matrix entry
+
+inputs:
+  endtoend:
+    type: boolean
+    default: false
+  endtoend_suite:
+    type: string
+    required: false
+    default: ''
+  endtoend_config:
+    type: string
+    required: false
+    default: ''
+  phpcoverage:
+    type: boolean
+    default: false
+  phplinting:
+    type: boolean
+    default: false
+  phpunit:
+    type: boolean
+    default: false
+  phpunit_suite:
+    type: string
+    required: false
+    default: ''
+  phpunit_fail_on_warning:
+    type: boolean
+    default: false
+  js:
+    type: boolean
+    default: false
+
+runs:
+  using: composite
+  steps:
+
+    - name: Validate inputs
```

Comment:
Does this step failing actually stop any of the other steps from running? If not, it's probably better for steps that use these inputs to run the validation, so that invalid inputs fail those steps.

Reply:
Single step failing fails the whole job and stops further steps, unless the step contains `if: always()`, hence why it's there for uploading artifacts

Note: unit/endtoend test step fails only after all the tests have run, then phpunit/behat will `exit 1`

Reply:
> Single step failing fails the whole job and stops further steps

Does this mean that if a phpunit test fails, it won't move on to do end-to-end tests?

Reply:
Does this mean that if a phpunit test fails, it won't move on to do end-to-end tests?

It would stop subsequent actions, but this isn't how jobs are setup to run by gha-generate-matrix which will only run a single type of test e.g. phpunit, endtoend, phplinting

> By "all the tests" here I assume you mean the respective tests - i.e. all unit tests will run before phpunit exits, but phpunit will not wait for end-to-end tests?

Steps do not run concurrently within a job.  If you did setup a job to run different test types (e.g. phpunit: true + endtoend: true) via the extra_jobs config, then the phpunit step would run first, it fails, then endtoend step would not run

Here's an example of matrix run to help clarify how it all works - https://github.com/emteknetnz/silverstripe-campaign-admin/actions/runs/2394020044

Reply:
Ahh yes, sorry. I keep getting confused between "job" in the sense of declaring a job in an action and "job" in the sense that the matrix array contains multiple "jobs".